src/Disks/ObjectStorages/MetadataStorageFromPlainRewritableObjectStorage.cpp

```diff
@@ -149,51 +168,100 @@ std::vector<std::string> getDirectChildrenOnRewritableDisk(
-                duplicates_filter.emplace(it->second);
-            /// The remote subdirectory name is the same as the local subdirectory.
-            else
-                duplicates_filter.emplace(path.substr(child_pos, slash_pos - child_pos));
+                result.emplace(std::move(filename));
         }
     }
-
-    return std::vector<std::string>(std::make_move_iterator(duplicates_filter.begin()), std::make_move_iterator(duplicates_filter.end()));
 }
 
 }
 
 MetadataStorageFromPlainRewritableObjectStorage::MetadataStorageFromPlainRewritableObjectStorage(
     ObjectStoragePtr object_storage_, String storage_path_prefix_)
     : MetadataStorageFromPlainObjectStorage(object_storage_, storage_path_prefix_)
-    , path_map(std::make_shared<PathMap>(loadPathPrefixMap(object_storage->getCommonKeyPrefix(), object_storage)))
+    , metadata_key_prefix(DB::getMetadataKeyPrefix(object_storage))
+    , path_map(std::make_shared<InMemoryPathMap>(loadPathPrefixMap(metadata_key_prefix, object_storage)))
 {
     if (object_storage->isWriteOnce())
         throw Exception(
             ErrorCodes::LOGICAL_ERROR,
             "MetadataStorageFromPlainRewritableObjectStorage is not compatible with write-once storage '{}'",
             object_storage->getName());
 
-    auto keys_gen = std::make_shared<CommonPathPrefixKeyGenerator>(object_storage->getCommonKeyPrefix(), metadata_mutex, path_map);
-    object_storage->setKeysGenerator(keys_gen);
+    if (getMetadataKeyPrefix() == object_storage->getCommonKeyPrefix())
+    {
+        auto keys_gen = std::make_shared<CommonPathPrefixKeyGenerator>(object_storage->getCommonKeyPrefix(), path_map);
+        object_storage->setKeysGenerator(keys_gen);
+    }
+    else
+    {
+        /// Use flat directory structure if the metadata is stored separately from the table data.
+        auto keys_gen = std::make_shared<FlatDirectoryStructureKeyGenerator>(object_storage->getCommonKeyPrefix(), path_map);
+        object_storage->setKeysGenerator(keys_gen);
+    }
 }
 
 MetadataStorageFromPlainRewritableObjectStorage::~MetadataStorageFromPlainRewritableObjectStorage()
 {
     auto metric = object_storage->getMetadataStorageMetrics().directory_map_size;
-    CurrentMetrics::sub(metric, path_map->size());
+    CurrentMetrics::sub(metric, path_map->map.size());
+}
+
+bool MetadataStorageFromPlainRewritableObjectStorage::exists(const std::string & path) const
+{
+    if (MetadataStorageFromPlainObjectStorage::exists(path))
+        return true;
+
+    if (getMetadataKeyPrefix() != object_storage->getCommonKeyPrefix())
+    {
+        auto key_prefix = object_storage->generateObjectKeyForPath(path, getMetadataKeyPrefix()).serialize();
+        return object_storage->existsOrHasAnyChild(key_prefix);
+    }
+
+    return false;
+}
+
+bool MetadataStorageFromPlainRewritableObjectStorage::isDirectory(const std::string & path) const
+{
+    if (getMetadataKeyPrefix() != object_storage->getCommonKeyPrefix())
```

Comment:
May be move `getMetadataKeyPrefix() != object_storage->getCommonKeyPrefix()` into separate method named `useSeparateLayoutForMetadata`? It is used quite often